state_machine/state_machine.hpp

```diff
@@ -0,0 +1,165 @@
+	/**
+	 * \brief           Makes a state which the state machine will use for execution.
+	 * 					DO NOT CALL THIS FUNCTION AND NOT PASS THE STATE TO THE STATE MACHINE
+	 * \param args		The arguments that will be passed to the constructor of the state
+	 */
+	template<typename T, typename ...Args>
+		static auto make_state(Args... args) -> T*{
+			static_assert(std::derived_from<T, State>, "State Must Be Derived From The State Class");
+			return new T(args...);
+		}
+
+	/**
+	 * \brief           Returns the name of the state machine
+	 * \return          A constant reference to the name of the state machine
+	 */
+	auto getName() const -> std::string const& {
+		return mName;
+	}
+
+	/**
+	 * \brief           Returns the demangled name of the state at runtime
+	 * \param state		A pointer to a state derived object which will have its runtime type analyzed
+	 * \return          A constant reference to the demangled state name at runtime
+	 */
+	auto getStateName(State const* state) const -> std::string const&{
+		return decoder.find(typeid(*state).hash_code())->second;
+	}
+
+	/**
+	 * \brief           Returns the demangled name of the current state in the state machine
+	 * \return          A constant reference to the current state's demangled state name
+	 */
+	auto getCurrentStateName() const -> std::string const& {
+		return getStateName(currState);
+	}
+
+	/**
+	 * \brief           Returns a map of type hashes to a vector of each type hash
+	 * \return          A constant reference to the map describing all valid state transitions
+	 */
+	auto getTransitionTable() const -> std::unordered_map<TypeHash, std::vector<TypeHash>> const&{
+		return mValidTransitions;
+	}
+
+	/**
+	 * \brief           Takes in a type hash and returns the demangled state name
+	 * \return          A constant reference to the demangled state name
+	 */
+	auto decodeTypeHash(TypeHash hash) const -> std::string const&{
+		return decoder.find(hash)->second;
+	}
+
+	/**
+	 * \brief           Enables the state transition from the first templated type to the subsequent templated types
+	 */
+	template<typename From, typename ...To>
+		void enableTransitions(){
+			static_assert(std::derived_from<From, State>, "From State Must Be Derived From The State Class");
+			static_assert((std::derived_from<To, State> && ...), "All States Must Be Derived From The State Class");
+			// Add From State To Decoder
+			int status = 0;
+			char* demangledName = abi::__cxa_demangle(typeid(From).name(), nullptr, nullptr, &status);
+
+			if(status){
+				throw std::runtime_error("C++ demangle failed!");
+			}
+
+			addNameToDecoder(typeid(From).hash_code(), demangledName);
+			free(demangledName);
+
+			mValidTransitions[typeid(From).hash_code()] = {typeid(To).hash_code()...};
+
+			std::vector<std::reference_wrapper<std::type_info const>> types{std::ref(typeid(To))...};
+			for(auto const& type : types){
+				demangledName = abi::__cxa_demangle(type.get().name(), nullptr, nullptr, &status);
```

Comment:
wtf

Reply:
its c++filt... but in C++ :100:

Reply:
This is unholy. Just have states carry a static member for the name...

Reply:
unholy is harsh... :cry: but it is p cool :sunglasses: